### PR TITLE
Code Insights: Add persistence to the code insights drill down filters

### PR DESCRIPTION
**Pull request by vovakulikov**
## Context

Closes https://github.com/sourcegraph/sourcegraph/issues/23683

This PR adds persistence logic to the filter panel of insight. By this, we cover a few cases
- You opened the filter panel on insight without filters. In this case, you can see the disabled save filter button (that is disabled until you change some values of the filter panel) 

Screenshot 2021-08-12 at 17 05 05

- User put some value in the filters panel and close this panel by outside click or escape
In this case, we show a chart with applied filters but do not store them (they are stored only in runtime and will be erased if the user reloads the page). We highlight that you have some filters in the filter panel by a small purple dot next to the filter icon.

Screenshot 2021-08-12 at 17 03 45

- If a user opened the filter panel and saves some filters they also will see a purple dot next to the filter icon. Next time when the user opens the filter panel they will see comments that some filters were applied and we also change the copy of the save button.

Screenshot 2021-08-12 at 17 09 49

## Technical notes

The drill-down filters are stored within the insight configuration object in the settings cascade. To add these filters properly we've created a few insight-specific JSON schema definitions to support filters in BE insights objects. 

Example of insight config with filters
```jsonc
"searchInsights.insight.soManyInsights": {
      "title": "So many insights",
      "series": [
        {
          "name": "\"insights\" insights",
          "stroke": "var(--oc-grape-7)",
          "query": "insights"
        }
      ],
      "filters": {
        "includeRepoRegexp": "",
        "excludeRepoRegexp": "sourcegrahp"
      }
    },
```

## Changes
- [x] Add save filters UI 
- [x] Change the insight types and add drill-down models to them. 
- [x] Integrate save and edit filters logic to the backend insight card

### CODE REVIEW

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrillDownFiltersForm.tsx [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+     * save/update default filters.
+     */
+    originalFiltersValue: DrillDownFiltersFormValues
+
+    /**
+     * Live filters that lives only in runtime memory and can be different
+     * from originalFiltersValue of insight until the user syncs them by
+     * save/update default filters.
+     */
+    initialFiltersValue: DrillDownFiltersFormValues
```
Comment (unclejustin) [inline client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrillDownFiltersForm.tsx:43]: **praise:** Thank you for these comments! Super helpful 🎉

**Review by unclejustin — APPROVED** [review]
Thanks @vovakulikov LGTM! 🌮 

I had a couple questions, but nothing blocking.

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/BackendInsight.tsx [hunk 2]
```diff
@@ -37,20 +39,51 @@ interface BackendInsightProps
...
+            const editedSettings = addInsightToSettings(settings.contents, insightWithNewFilters)
+
+            await updateSubjectSettings(platformContext, subjectId, editedSettings).toPromise()
+
+            telemetryService.log('CodeInsightsSearchBasedFilterUpdatingClick')
+
+            setOriginalInsightFilters(filters)
+            setIsFiltersOpen(false)
+        } catch (error) {
+            return { [FORM_ERROR]: asError(error) }
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply client/web/src/insights/components/insights-view-grid/components/backend-insight/BackendInsight.tsx:80]: @unclejustin Yeah. The filter panel has UI to show error messages.

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrillDownFiltersForm.tsx (hunk 1 above)
Comment (valerybugakov) [reply client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrillDownFiltersForm.tsx:43]: ➕

**Comment by valerybugakov** [comment]
> User put some value in the filters panel and close this panel by outside click or escape
In this case, we show a chart with applied filters but do not store them (they are stored only in runtime and will be erased if the user reloads the page)

UX feedback: I would expect the closed filter panel to behave consistently (if the filter is applied —> then it should always be persisted) because there's no explanation that if the panel is closed by an outside click, the applied filter is different from the saved one. Another possibility to make it clearer: don't apply a filter after an outside click. 

I might misunderstand the use case, but it seems like an introduction of the state hidden from the user, which can be confusing.

File: client/web/src/insights/components/insights-view-grid/SmartInsightsViewGrid.tsx [hunk 3]
```diff
@@ -37,7 +39,19 @@ export const SmartInsightsViewGrid: React.FunctionComponent<SmartInsightsViewGri
...
                 />
-            ))}
+            )),
+        // Ignore settings cascade change in order to avoid grid item re-render and
+        // grid position items animations. In some cases (like insight filters updating
+        // we want to ignore changes of insights from settings cascade).
+        // But still trigger grid animation rerender if insight ordering or insight count
+        // have been changed.
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+        [telemetryService, platformContext, extensionsController, insightIds]
```
[the remark this answers is not in the document]
Comment (valerybugakov) [reply client/web/src/insights/components/insights-view-grid/SmartInsightsViewGrid.tsx:49]: Do I understand correctly that we don't want to re-render `SmartInsight` on the `settingsCascade` change? 

If that's the only objective of the `useMemo` call, it might be a good idea to consider wrapping `SmartInsight` into `React.memo` call where we could clearly indicate that we don't care about updates of one specific prop — `settingsCascade`. It would clear from this logic what we're trying to do without comment, and re-renders of `insights` on order and count change will stay in place.

```tsx
export const SmartInsight = React.memo(SmartInsight, shallowEqualsCheckIgnoringSettingsCascade)
```

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/DrillDownFiltersPanel.tsx [hunk 4]
```diff
@@ -1,141 +1,58 @@
...
-                error={includeRegex.meta.dirty && includeRegex.meta.error}
-                {...includeRegex.input}
+    const handleFilterSave = (values: DrillDownFiltersFormValues): SubmissionResult =>
+        onFilterSave({
+            includeRepoRegexp: values.includeRepoRegexp,
+            excludeRepoRegexp: values.excludeRepoRegexp,
+        })
+
+    // By default always render filters mode
+    const [step] = useState(DrillDownFiltersStep.Filters)
```
Comment (valerybugakov) [inline client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/DrillDownFiltersPanel.tsx:43]: To confirm: `ViewCreation` toggle will be implemented in a separate PR?

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrillDownFiltersForm.tsx [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+export const DrillDownFiltersForm: React.FunctionComponent<DrillDownFiltersFormProps> = props => {
+    const { className, initialFiltersValue, originalFiltersValue, onFiltersChange, onFilterSave } = props
+
+    const { ref, formAPI, handleSubmit, values } = useForm<DrillDownFiltersFormValues>({
+        initialValues: initialFiltersValue,
+        onChange: onFiltersChange,
+        onSubmit: onFilterSave,
+    })
```
Comment (valerybugakov) [inline client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrillDownFiltersForm.tsx:63]: It took me some time to understand why we need to save the same state in two places: here and in `BackendInsight.tsx`. 

Would it be possible to have only one source of state if we would have `FormContext`? Meaning: create form in `BackendInsight.tsx` and then `useField` in `DrillDownFiltersForm` using `formAPI` from `FormContext`.

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/components/drill-down-reg-exp-input/DrillDownRegExpInput.module.scss [hunk 6]
```diff
@@ -0,0 +1,23 @@
+.regexp-field {
+    display: flex;
+
+    &__prefix {
```
Comment (valerybugakov) [inline client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/components/drill-down-reg-exp-input/DrillDownRegExpInput.module.scss:4]: > Do not use BEM convention in CSS modules. Use short descriptive classes specific only for the corresponding component because CSS modules provide scoping out of the box. It outputs shorter classes that are more readable in the component markup.

As we agreed to do from our [docs](https://docs.sourcegraph.com/dev/background-information/web/styling#css-modules).

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/components/drill-down-reg-exp-input/DrillDownRegExpInput.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+export const LabelWithReset: React.FunctionComponent<PropsWithChildren<LabelWithResetProps>> = props => (
+    <span className="d-flex align-items-center">
+        <TruncatedText>{props.children}</TruncatedText>
+        <button
+            type="button"
+            className="btn btn-link ml-auto pt-0 pb-0 pr-0 font-weight-normal"
+            onClick={props.onReset}
+        >
+            Reset
+        </button>
```
Comment (valerybugakov) [inline client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/components/drill-down-reg-exp-input/DrillDownRegExpInput.tsx:43]: Consider using components from the Wildcard package if they are available for new interfaces, and please report issues if it's inconvenient to use them for some reason. 

For example, [the Button component](https://github.com/sourcegraph/sourcegraph/blob/main/client/wildcard/src/components/Button/Button.tsx) can be used here.

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrilDownFiltersForm.story.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    <WebStory>{() => story()}</WebStory>
+))
+
+const fakeAPIRequest = async () => {
+    await delay(1000)
+
+    return { [FORM_ERROR]: new Error('Fake api request error') }
+}
+
+const EMPTY_DRILLDOWN_FILTERS: DrillDownFiltersFormValues = {
```
Comment (valerybugakov) [inline client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrilDownFiltersForm.story.tsx:20]: Is it possible to reuse `EMPTY_DRILLDOWN_FILTERS ` from `drill-down-filters-panel/utils.ts`?

File: client/web/src/integration/insights/insights-visual.test.ts [hunk 9]
```diff
@@ -69,7 +69,9 @@ describe('[VISUAL] Code insights page', () => {
         await takeChartSnapshot('Code insights page with back-end insights only')
     })
 
-    it('is styled correctly with search-based insights ', async () => {
+    // Unmute that test when flaky issue with line chart tooltip will be resolved
+    // see https://github.com/sourcegraph/sourcegraph/issues/23669
```
Comment (valerybugakov) [inline client/web/src/integration/insights/insights-visual.test.ts:73]: 🙌

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/components/drill-down-reg-exp-input/DrillDownRegExpInput.module.scss [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        font-family: monospace;
+        flex-shrink: 0;
+        padding: 0.25rem 0.5rem;
+        height: 100%;
+        border: 1px solid var(--input-border-color);
+        background: var(--input-border-color);
+        border-top-left-radius: 4px;
+        border-bottom-left-radius: 4px;
```
Comment (valerybugakov) [inline client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/components/drill-down-reg-exp-input/DrillDownRegExpInput.module.scss:12]: Is it possible to use `--border-radius` here to be consistent with the design system? 
If it's a one-off solution that the design team is aware of, consider commenting on why it's necessary here.

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrilDownFiltersForm.story.tsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+const { add } = storiesOf('web/insights/DrillDownFilters', module).addDecorator(story => (
+    <WebStory>{() => story()}</WebStory>
+))
+
+const fakeAPIRequest = async () => {
+    await delay(1000)
+
+    return { [FORM_ERROR]: new Error('Fake api request error') }
+}
```
Comment (valerybugakov) [inline client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrilDownFiltersForm.story.tsx:18]: This helper looks like great general-purpose util to be resued in other tests where forms are involved.

**Review by valerybugakov — APPROVED** [review]
Great work, @vovakulikov!

Annoying comment: the PR is quite big comparing to [our recommended changeset size](https://about.sourcegraph.com/handbook/engineering/developer-insights#prefer-small-prs-lines). Consider pulling some independent bits out in a separate pull request next time.

I struggle with it myself, but it's a breeze for reviewers to pick up small PRs for review. Let's keep pinging each other about that.  🙂

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/DrillDownFiltersPanel.tsx (hunk 4 above)
Comment (vovakulikov) [reply client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/DrillDownFiltersPanel.tsx:43]: This is correct. Here is the ticket for this functionality https://github.com/sourcegraph/sourcegraph/issues/23233

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrillDownFiltersForm.tsx (hunk 5 above)
Comment (vovakulikov) [reply client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrillDownFiltersForm.tsx:63]: Yeah, I also was thinking about this. But decided to stick with the current implementation (state uncontrolled form component) in order to keep this PR small (but even with that it's 800 loc) 
I created an issue to cover this case here https://github.com/sourcegraph/sourcegraph/issues/23890

Thanks for suggestion

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/components/drill-down-reg-exp-input/DrillDownRegExpInput.module.scss (hunk 6 above)
Comment (vovakulikov) [reply client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/components/drill-down-reg-exp-input/DrillDownRegExpInput.module.scss:4]: Ah, missed that again. Thanks, I will move them in separate classes without nesting

**Comment by vovakulikov** [comment]
@valerybugakov thanks for your review. I think you're right about UX concerns and I'm with you about this. We've been discussing this behavior for a while already and we've got a couple of thoughts on this

- We can separate these two cases (saved filters and temporal filters by color coding) here is the designs and comments for context https://www.figma.com/file/1zvF8Owrh1ndVH29lvyrXn/RFC-419%3A-Code-Insights-“Drilldown”?node-id=1072%3A7302
- We also introduced something called default filters which can give some signs about how it works 
- We can stick with the current implementation and check users' feedback about this.

cc @AlicjaSuska @felixfbecker Valery concerns is a good example that current logic can be confused for users. I think we can stick with the current implementation for the beta but we want/should revisit this problem later. Also want to get your attention to this comment https://www.figma.com/file/1zvF8Owrh1ndVH29lvyrXn?node-id=1231:11178#100963083

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/components/drill-down-reg-exp-input/DrillDownRegExpInput.tsx (hunk 7 above)
Comment (vovakulikov) [reply client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/components/drill-down-reg-exp-input/DrillDownRegExpInput.tsx:43]: Cool, I've missed that we have a separate shared component now. I will check this and change this component.

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrilDownFiltersForm.story.tsx (hunk 8 above)
Comment (vovakulikov) [reply client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrilDownFiltersForm.story.tsx:20]: Just because the drill-down form is a child component of the filter panel I would avoid this re-use here. Cause usually some mental cyclical dependency makes this deps tree a bit complex than it should be. My second thought on that was maybe we should put down the `EMPTY_DRILLDOWN_FILTERS` on a lower level but we should consider that we FormValues Model (`DrillDownFiltersFormValues`) and filters model (`SearchBasedBackendFilters`) are two different levels. 

So technically we could re-use them but from a mental perspective, I wouldn't do that

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrilDownFiltersForm.story.tsx (hunk 11 above)
Comment (vovakulikov) [reply client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrilDownFiltersForm.story.tsx:18]: Agree, I also have some issue about a related thing. Could be part of that https://github.com/sourcegraph/sourcegraph/issues/21907

I will add this comment to the issue description.

File: client/web/src/insights/components/insights-view-grid/SmartInsightsViewGrid.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (vovakulikov) [reply client/web/src/insights/components/insights-view-grid/SmartInsightsViewGrid.tsx:49]: @unclejustin sure we can discuss this. 
@valerybugakov Yeah, you're right about this approach with the memo func, probably this approach is a bit better cause I wouldn't  need to violate eslint rule about hook dependencies 

Thanks!

File: client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrillDownFiltersForm.tsx (hunk 5 above)
Comment (valerybugakov) [reply client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrillDownFiltersForm.tsx:63]: Yeah, I agree with you here. I was curious to know your thoughts on that.

Comment (vovakulikov) [reply client/web/src/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/components/drill-down-filters-form/DrillDownFiltersForm.tsx:63]: The problem with that approach as I understand there is no way to achieve type safety here with sharing state via react context. In other words, an can't be sure at some form component that parent will pass a proper value with the right share as a value for the form. 

I think I'll end up with runtime check invariant about form shape and cast form API or throw assertion error otherwise.

**Review by felixfbecker — COMMENTED** [review]
Not urgent comments (not needed to fix before Beta)

File: schema/settings.schema.json [hunk 12]
```diff
@@ -521,6 +541,55 @@
...
+          "description": "Performs a search query and shows the number of results returned."
+        },
+        "search": {
+          "type": "string",
+          "description": "Performs a search query and shows the number of results returned."
+        },
+        "webhook": {
+          "type": "string",
+          "description": "(not yet supported) Fetch data from a webhook URL."
+        }
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply schema/settings.schema.json:590]: This diff is just a shirt of old BE insight schema which we are not ready to remove. It is actually schema from the first version of the prototype. It will be completely removed at some point after beta and before GA.

File: schema/settings.schema.json [hunk 13]
```diff
@@ -521,6 +541,55 @@
...
+          "type": "string",
+          "description": "A globally  unique identifier for this insight."
+        }
+      }
+    },
+    "InsightSeries": {
+      "type": "object",
+      "additionalProperties": false,
+      "required": ["label"],
+      "properties": {
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply schema/settings.schema.json:574]: Yes! You're absolutely right and I thought this way as well. But I spent a lot of time trying to figure out how to achieve union types in GO and there is no way to do that, at least I could find anything to do that. So because of that, I ended up with a simple properties map. I agree it should be union type with some type filters property (enum value)
